Tests Drawing Codecs (based on Mono's test suite)

qmfrederik opened the pull request:
Adds tests for the Bitmap, PNG, TIFF, JPEG, GIF and Icon codecs, by opening/saving various files in this format and verifying pixel data.

These tests validate both System.Drawing.Common as well as GDI+/libgdiplus.

qmfrederik commented:
/cc @mellinoe @safern @hughbe

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/TiffCodecTests.cs:
```diff
@@ -0,0 +1,323 @@
+            {
+                GraphicsUnit unit = GraphicsUnit.World;
+                RectangleF rect = bmp.GetBounds(ref unit);
+                // MS reports 24 bpp while we report 32 bpp
```
Ideally, this assertion would be in a separate test case, and disabled for Unix (against a descriptive issue). That way, the behavior difference is captured more succinctly and clearly than in a commented-out part of another test.

qmfrederik commented:
@mellinoe I think this one is good to go.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -0,0 +1,2005 @@
+                Assert.Equal(PixelFormat.Format32bppArgb, bmp.PixelFormat);
+                Assert.Equal(73746, bmp.Flags);
+
+                // These values are inconsistent accross Windows & Unix: 0 on Windows, 16 on Unix
```
We shouldn't leave commented-out code like this in the tests. If we are able to, we can move it into a Windows-specific test case, and disable it on Unix. If we don't do that, we should just remove the dead code.

qmfrederik replied:
@mellinoe I created two test cases, one for the Windows behavior and one for the Unix behavior. I did it because the Unix behavior is in a sense more complete: Unix returns palette data whereas Windows does not.

It also ensures getting the Palette will keep on working on Unix (as opposed to disabling the test altogether so that the code will never be exercised).

qmfrederik commented:
@dotnet-bot test Windows x86 Release Build

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -86,6 +83,30 @@ public void Image16()
             }
         }
 
+        [ActiveIssue(20844, TestPlatforms.Windows)]
```
We should use `[PlatformSpecific(TestPlatforms.AnyUnix)]` here.

mellinoe commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -86,6 +83,30 @@ public void Image16()
+            }
+        }
+
+        [ActiveIssue(20844, TestPlatforms.AnyUnix)]
```
We should use `[PlatformSpecific(TestPlatforms.Windows)]` here.

mellinoe commented:
@qmfrederik We should use the `PlatformSpecific` attribute for the new split tests, since they validate stuff for specific platforms.

qmfrederik commented:
@mellinoe OK, done!

mellinoe commented:
@qmfrederik Sorry for all the seemingly nitpicky problems 😄. Looks good to me now!

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/BmpCodecTests.cs:
```diff
@@ -0,0 +1,565 @@
+            string s;
+
+            int p = (int)Environment.OSVersion.Platform;
+            if ((p == 4) || (p == 128) || (p == 6))
```
I think this can be achieved with `PlatformDetection.IsWindows` property

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/GifCodecTests.cs:
```diff
@@ -0,0 +1,291 @@
+            string s;
+
+            int p = (int)Environment.OSVersion.Platform;
+            if ((p == 4) || (p == 128) || (p == 6))
```
Use `PlatformDetection.IsWindows`

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/BmpCodecTests.cs:
```diff
@@ -0,0 +1,565 @@
+                    Assert.Equal(PixelFormat.Format24bppRgb, data.PixelFormat);
+                    Assert.Equal(520, data.Stride);
+                    Assert.Equal(183, data.Height);
+                    int size = data.Height * data.Stride;
```
I think size is not used, you can delete this variable

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/BmpCodecTests.cs:
```diff
@@ -0,0 +1,565 @@
+
+        private void Save(PixelFormat original, PixelFormat expected, bool colorCheck)
+        {
+            string sOutFile = String.Format("linerect{0}-{1}.bmp", GetOutSufix(), expected.ToString());
```
nit: String => string

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/GifCodecTests.cs:
```diff
@@ -0,0 +1,291 @@
+                    Assert.Equal(PixelFormat.Format24bppRgb, data.PixelFormat);
+                    Assert.Equal(332, data.Stride);
+                    Assert.Equal(100, data.Height);
+                    int size = data.Height * data.Stride;
```
size is not used, can probably be deleted.

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -0,0 +1,2131 @@
+            string s;
+
+            int p = (int)Environment.OSVersion.Platform;
+            if ((p == 4) || (p == 128) || (p == 6))
```
Use `PlatformDetection.IsWindows`. 

Btw I see this method been used in multiple (or all) the test classes you're adding. Can we move it to a helper method?

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -0,0 +1,2131 @@
+            }
+        }
+
+        [PlatformSpecific(TestPlatforms.Windows]
```
You're missing a `)` here. This will make CI to fail

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -0,0 +1,2131 @@
+                    Assert.Equal(PixelFormat.Format24bppRgb, data.PixelFormat);
+                    Assert.Equal(32, data.Height);
+
+                    int size = data.Height * data.Stride;
```
size is not used.

safern commented on src/System.Drawing.Common/tests/mono/System.Drawing.Imaging/IconCodecTests.cs:
```diff
@@ -0,0 +1,2131 @@
+                    Assert.Equal(PixelFormat.Format24bppRgb, data.PixelFormat);
+                    Assert.Equal(48, data.Height);
+
+                    int size = data.Height * data.Stride;
```
size not used. I wonder if you're doing this to cause a property in data to be initialized.

qmfrederik commented:
@safern @mellinoe Thanks for the feedback! I updated the code, hope we're good this time :-).

Regarding the GetOutSuffix methods - turns out they are not really needed, so I just removed them.

Regarding the style-related nits, I was wondering why no Roslyn analyzers are enabled for corefx? We've saved ourselves lots of back-and-forth over nits in our PR reviews by enabling them and causing the build to fail on warnings. Plus, you get those Fix-Its for free :-)

mellinoe commented:
> Regarding the style-related nits, I was wondering why no Roslyn analyzers are enabled for corefx? We've saved ourselves lots of back-and-forth over nits in our PR reviews by enabling them and causing the build to fail on warnings. Plus, you get those Fix-Its for free :-)

We've thought about it, but never made anything mandatory. To be honest, there are a lot more things in this mono code that are non-standard for corefx, but I'm trying not to fix every single style issue given how much work it was to get to this point, and because of the fact that these are "legacy compat" tests in the first place. When we brought over the .NET Framework tests initially, they were in a similar boat.

qmfrederik commented:
@dotnet-bot test Linux arm Release Build 
@dotnet-bot test Tizen armel Debug Build

mellinoe commented:
@qmfrederik Thanks for all of the cleanup! Merging now.
